# WEEK -2 My notes

## The Series Data Structure

**Doubts:**

-What is Vectorization?

The series is one of the core data structures in pandas. You think of it a cross between a list and a dictionary.The items are all stored in an order and there's labels with which you can retrieve them. An easy way to visualize this is two columns of data. The first is the special index, a lot like the dictionary keys. While the second is your actual data. It's important to note that the data column has a label of its own and can be retrieved using the .name attribute.  
  
The name attribute usage: https://stackoverflow.com/questions/31292140/what-is-the-name-parameter-in-pandas-series

In [5]:
import pandas as pd
import numpy as np
pd.Series

pandas.core.series.Series

In [3]:
animals=["Tiger","Lion","Zebra","Monkey"]
animals_ser=pd.Series(animals,name="Animals")
print(animals_ser)
animals_df=pd.DataFrame(animals_ser)
animals_df

0     Tiger
1      Lion
2     Zebra
3    Monkey
Name: Animals, dtype: object


Animals
0   Tiger
1    Lion
2   Zebra
3  Monkey

In [4]:
numbers=[1,2,3,4]
pd.Series(numbers,name="Numbers")

0    1
1    2
2    3
3    4
Name: Numbers, dtype: int64

When there is lack of datapoint, it is considered as None objectType in case of list of Strings

In [5]:
animals=["Tiger","Lion","Zebra",None]
pd.Series(animals)

0    Tiger
1     Lion
2    Zebra
3     None
dtype: object

When there is lack of datapoint, it is considered as Nan floatType in case of list of integers

In [6]:
numbers=[1,2,3,None]
pd.Series(numbers)

0    1.0
1    2.0
2    3.0
3    NaN
dtype: float64

Nan and None are not the same.They are not equal in terms of Numerical value.

In [7]:
import numpy as np
np.nan == None

False

In [8]:
np.isnan(np.nan)

True

In [9]:
alphabets={'A':'a','B':'b',
          'C':'c','D':'d'}
s= pd.Series(alphabets)
s

A    a
B    b
C    c
D    d
dtype: object

In [10]:
s.index

Index(['A', 'B', 'C', 'D'], dtype='object')

### Querying a Series

A panda.Series can be queried, either by the index position or the index label.

In [11]:
alphabets={'A':'a','B':'b',
          'C':'c','D':'d'}
s= pd.Series(alphabets)
s

A    a
B    b
C    c
D    d
dtype: object

Query using index number

In [11]:
s.iloc[3]

'd'

Query using index name

In [12]:
s.loc['D']

'd'

***Note:*** When the index is list of numbers, pandas cant determine whether u r intending to query by index position or index lable.

### What else we can do with Series?
Okay, so now we know how to get data out of the series. Let's talk about working with the data.  
A common task is to want to consider all of the values inside of a series and want to do some sort of operation. This could be trying to find a certain number, summarizing data or transforming the data in some way.  

In [13]:
s=pd.Series([100.00,120.00,101.00,3.00])
s

0    100.0
1    120.0
2    101.0
3      3.0
dtype: float64

A typical programmatic approach to this would be to iterate over all the items in the series, and invoke the operation one is interested in. For instance, we could create a data frame of floating point values. Let's think of these as prices for different products. We could write a little routine which iterates over all of the items in the series and adds them together to get a total.

In [14]:
#summing without vectorization
total=0
for item in s:
    total+=item
print(total)

324.0


 Let's think of these as prices for different products. We could write a little routine which iterates over all of the items in the series and adds them together to get a total.
  
This works, but it's slow. Modern computers can do many tasks simultaneously, especially, but not only, tasks involving mathematics.
**Pandas and the underlying NumPy libraries support a method of computation called vectorization.**

In [16]:
import numpy as np
#summing using vectorization
total=np.sum(s)
print(total)


324.0


**Now both of these methods create the same value, but is one actually faster?**


In [17]:
#Let's create a big series of random numbers
s=pd.Series(np.random.randint(0,1000,10000))
s

0       990
1       422
2       209
3       166
4       238
       ... 
9995    284
9996    681
9997    225
9998    129
9999    330
Length: 10000, dtype: int64

The Jupyter Notebook has a magic function which can help to compute the time taken by each bunch of code to execute.  
Magic functions begin with a percentage sign. If we type this sign and then hit the Tab key, we can see a list of the available magic functions. You could write your own magic functions too, but that's a little bit outside of the scope of this course. We're actually going to use what's called a **cellular magic function.**  
  
The function is called **timeit**

In [18]:
%%timeit -n 100
#Without Vectorization
summary=0
for item in s:
    summary+=item
    

2.21 ms ± 111 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [19]:
%%timeit -n 100
#With Vectorization
summary=np.sum(s)

146 µs ± 14.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


This is a pretty shocking difference in the speed and demonstrates why data scientists need to be aware of parallel computing features and start thinking in functional programming terms.

**Related feature in Pandas and NumPy is called broadcasting.** With broadcasting, you can apply an operation to every value in the series, changing the series.

For instance, if we wanted to increase every random variable by 2, we could do so quickly using the += operator directly on the series object. Here I'll just use the head operator to just print out the top five rows in the series.  
  
The procedural way of doing this would be to iterate through all of the items in the series and increase the values directly. A quick aside here. Pandas does support iterating through a series much like a dictionary, allowing you to unpack values easily. But if you find yourself iterating through a series, you should question whether you're doing things in the best possible way.

In [22]:
%%timeit -n 1
s=pd.Series(np.random.randint(0,1000,10000))
for label,value in s.iteritems():
    s.loc[label]=value+2
    print(s.loc[label])

696
354
978
687
471
741
421
952
670
764
512
401
726
559
347
466
103
71
786
598
259
647
928
159
149
678
2
557
947
27
118
109
753
810
470
105
309
43
513
655
69
809
987
710
371
355
731
647
878
533
147
335
220
297
657
367
509
370
232
284
908
759
563
328
629
167
231
838
628
968
569
47
739
672
120
795
686
908
457
503
561
529
675
282
978
690
154
296
902
586
70
474
708
901
254
868
678
122
172
504
939
654
899
584
13
533
952
271
85
793
259
989
59
176
658
464
735
684
387
977
674
42
564
513
829
554
560
428
361
250
400
991
308
625
799
475
737
444
719
461
410
391
318
306
733
918
363
434
542
729
18
597
566
554
344
5
935
728
824
758
200
152
856
457
878
102
291
847
202
994
129
92
419
146
974
401
544
661
849
372
716
296
458
807
739
555
273
840
988
221
918
113
90
218
639
222
476
927
480
556
55
741
636
76
751
287
630
673
651
555
980
907
392
748
661
83
480
134
682
901
813
794
351
809
543
703
86
471
305
613
510
138
484
965
594
662
554
654
884
959
841
929
924
224
550
110
948
913
651
383
452
65
712
619
494
79

58
812
291
645
76
9
541
331
531
325
955
674
138
125
853
929
288
69
546
83
971
893
659
987
468
992
648
69
368
738
832
939
673
357
606
857
17
136
939
965
183
485
320
32
800
638
214
913
714
78
969
975
799
169
289
215
920
148
422
784
12
557
771
520
529
654
623
477
42
83
678
225
782
494
856
319
668
965
189
141
469
97
771
640
444
819
188
99
556
974
1000
899
413
995
545
640
129
467
723
911
968
343
209
584
488
602
583
337
234
312
571
610
144
752
265
533
780
367
30
367
541
904
757
590
34
175
881
278
915
327
529
581
148
940
379
473
463
254
402
156
751
434
853
45
673
303
720
955
503
981
951
970
118
600
208
317
656
144
533
341
632
584
974
812
966
960
81
433
442
319
769
192
664
112
829
470
314
198
648
377
887
502
442
372
973
283
719
362
631
300
536
644
940
535
903
168
56
790
55
634
802
184
811
491
688
904
662
465
544
687
417
50
894
130
445
535
856
997
246
717
834
118
156
837
811
370
130
747
574
26
804
831
660
1000
451
183
170
5
740
829
102
275
697
657
160
963
931
453
9
701
740
494
793
558
837
941
3

955
870
368
340
664
111
305
416
877
99
952
625
132
782
489
528
590
866
973
57
157
453
239
226
27
899
699
177
662
625
395
686
152
75
896
98
999
673
528
444
820
283
711
696
387
718
454
111
948
270
250
471
976
705
256
394
311
346
473
497
739
992
222
173
608
69
170
994
382
598
84
464
832
746
629
873
382
570
318
697
94
560
650
136
424
50
929
82
360
171
562
394
361
365
209
908
27
757
267
889
254
715
657
685
237
577
978
966
26
21
653
192
150
943
333
470
144
262
955
311
912
692
357
314
554
263
644
957
822
748
909
489
348
497
391
828
460
268
351
116
74
148
177
415
619
159
439
505
343
777
506
105
413
389
480
933
274
825
562
710
851
755
776
965
953
675
723
488
422
371
785
321
569
223
785
369
652
395
332
381
682
856
531
991
812
767
61
971
944
850
723
508
91
465
909
683
795
363
623
104
364
533
43
108
300
547
990
589
530
511
740
498
724
533
808
36
833
841
852
132
790
547
338
96
800
633
594
488
914
53
69
478
765
403
500
855
339
128
18
172
535
28
892
588
465
970
342
576
554
566
169
701
972
104
147
950

371
921
87
211
794
300
344
467
469
564
983
648
968
747
863
598
863
91
279
10
902
32
770
128
472
570
653
161
483
509
568
541
921
271
375
144
820
574
613
672
898
402
591
249
869
115
137
682
429
159
313
625
330
401
62
501
447
547
719
239
499
41
98
928
660
627
857
387
462
16
249
687
980
757
465
382
230
553
248
280
532
984
527
86
660
920
669
143
610
968
975
119
792
577
923
988
917
934
20
532
717
783
320
182
870
94
814
466
55
697
153
895
342
528
519
343
707
169
494
531
330
267
939
413
861
275
688
277
13
143
651
96
579
236
422
937
289
303
814
680
326
839
662
971
67
66
970
850
355
874
788
159
108
510
471
528
346
198
675
144
798
530
498
47
8
22
686
623
462
184
703
238
204
255
461
989
577
597
931
302
841
405
855
956
158
568
259
474
247
347
960
944
771
52
762
255
759
772
438
135
702
582
511
202
961
498
262
343
251
553
936
912
895
493
835
104
473
26
862
877
102
258
585
448
538
927
351
960
541
165
101
892
43
122
712
443
733
746
157
845
443
884
326
804
35
174
350
536
50
719
556
626
33
636
236
781
92

61
389
552
831
763
22
472
528
220
751
342
973
812
361
594
16
544
36
345
490
888
809
795
568
504
679
877
382
846
256
854
720
351
36
192
478
6
782
458
157
429
619
576
520
770
366
54
599
370
91
319
453
447
43
119
250
912
744
107
741
814
714
462
111
667
330
941
171
748
607
862
118
728
913
630
439
434
768
7
857
496
519
111
74
692
109
153
686
33
506
929
600
138
919
69
723
540
113
3
783
5
308
9
921
214
778
202
430
371
520
906
888
645
780
719
5
188
698
34
679
927
737
994
96
803
749
640
925
216
267
848
721
729
908
5
935
600
402
747
291
932
953
264
471
353
402
224
473
722
982
882
892
382
265
428
329
423
584
17
186
560
603
813
647
695
582
722
625
267
983
484
298
943
774
509
442
224
491
943
674
243
792
482
629
66
440
233
278
544
509
159
192
210
424
599
104
998
902
268
49
15
201
694
631
155
219
860
603
154
281
909
938
891
527
681
759
735
432
62
748
35
973
306
414
55
383
83
254
354
530
505
525
892
842
573
242
987
915
138
197
138
240
587
541
725
709
309
308
456
752
7
839
357
472
711
224
245
161
964
9

689
336
365
293
47
239
832
150
259
279
166
845
337
791
123
740
247
691
483
556
639
208
199
6
64
571
125
228
306
856
674
206
241
745
711
65
146
711
497
311
845
490
384
399
845
689
986
593
719
883
119
4
794
252
201
127
278
755
845
883
292
386
688
130
938
432
446
912
691
571
687
65
910
586
567
386
219
335
46
191
849
483
543
699
718
122
767
46
792
77
310
438
866
705
976
948
294
77
198
359
381
818
614
158
62
77
683
28
899
29
540
624
646
370
8
217
940
147
54
982
481
454
652
302
177
963
539
352
862
586
403
658
304
105
611
107
23
855
403
614
88
967
686
515
579
355
524
465
33
53
490
40
404
67
287
991
382
248
632
523
427
987
864
977
794
494
456
440
433
257
354
829
1001
862
717
73
843
73
168
522
771
78
961
373
631
902
376
853
383
685
623
468
83
559
610
596
107
996
673
636
166
958
989
402
912
356
35
657
800
1000
829
513
998
720
85
812
702
335
641
52
639
309
619
349
340
659
302
181
385
499
321
524
10
516
563
266
859
477
86
240
125
967
188
593
125
26
552
121
66
167
867
145
263
306
18
78
303
907
790


828
919
82
144
123
835
641
337
194
964
593
141
899
366
27
596
381
676
487
579
941
676
809
968
581
616
546
977
494
588
169
874
640
548
731
592
216
117
370
4
736
333
725
484
545
928
893
879
730
673
160
760
589
998
368
804
456
263
274
434
609
437
147
75
660
504
716
826
673
788
602
737
608
516
148
666
221
954
756
571
828
471
178
261
415
291
839
861
573
150
246
355
18
155
404
443
128
974
409
645
541
18
756
96
268
562
842
585
219
391
100
128
520
551
16
375
688
125
876
369
847
813
142
589
576
737
135
208
770
755
142
93
647
92
265
650
490
884
867
630
479
432
510
153
989
206
697
724
853
890
689
324
838
253
931
682
464
226
351
135
460
740
343
964
710
783
286
796
624
549
204
342
85
620
140
704
550
70
323
438
27
820
489
56
844
387
768
226
740
931
254
304
537
999
388
168
906
552
193
627
663
798
345
546
303
168
584
88
699
113
332
809
145
361
934
242
152
954
353
300
886
382
796
331
908
193
721
232
37
374
457
566
509
874
414
440
361
959
21
113
442
16
669
62
170
647
228
207
831
920
628
563
569
896
140


709
172
961
573
680
646
108
629
364
552
262
453
64
194
228
54
900
996
828
832
154
217
77
750
816
546
794
259
211
678
187
670
111
38
171
679
246
743
795
845
876
154
753
447
6
777
441
256
745
232
680
453
773
140
994
75
777
2
172
805
770
708
78
833
810
23
722
608
451
298
706
561
233
224
975
28
965
328
772
421
382
294
697
702
818
699
25
236
2
367
96
429
389
897
49
765
411
258
588
503
856
781
213
337
189
791
288
858
337
967
964
85
491
219
226
761
133
371
921
463
222
299
267
818
940
718
487
181
442
846
627
348
504
212
95
609
54
3
609
676
394
830
619
560
278
237
227
77
378
746
439
537
986
120
938
866
668
439
787
661
933
328
580
690
859
201
583
320
369
455
306
539
228
189
602
628
274
400
798
731
930
737
363
142
295
563
743
345
783
235
222
61
69
351
134
891
474
628
643
788
915
772
129
1000
702
490
305
953
316
348
532
40
719
511
448
927
991
25
337
148
649
755
67
552
351
361
478
979
187
275
314
841
498
22
263
184
544
245
460
378
161
503
9
656
754
458
39
526
93
406
157
383
562
766
845
756
998
565


168
346
299
50
128
154
21
789
697
315
897
742
196
21
995
855
587
236
812
421
465
847
439
263
610
954
432
294
287
958
253
452
9
885
766
135
100
647
418
245
351
891
447
495
764
186
474
880
732
211
66
648
736
860
210
468
778
731
420
922
513
895
431
993
774
313
439
793
809
817
835
944
399
849
789
976
359
687
48
131
498
264
714
933
125
391
339
202
440
754
826
277
182
633
735
102
280
47
980
459
863
916
874
628
19
503
694
501
582
390
427
212
465
541
36
73
380
622
544
473
885
405
857
715
507
766
578
774
758
271
203
182
806
911
147
963
777
523
709
84
854
569
112
842
367
24
376
364
289
233
429
271
317
671
728
952
942
219
875
17
392
887
785
821
436
773
774
703
355
998
300
753
923
514
109
161
700
773
133
747
641
85
910
216
851
781
70
529
285
398
57
119
899
643
264
224
694
717
23
25
537
578
309
183
302
137
875
667
717
578
368
696
593
155
644
241
99
495
813
275
471
152
989
388
297
729
393
758
645
281
476
846
175
276
446
991
10
319
164
603
249
430
186
252
216
786
19
470
304
744
304
524
394
149
541
79

989
434
41
499
669
639
537
928
807
421
876
81
177
985
642
397
677
145
220
179
918
271
923
562
729
817
278
395
68
656
793
154
180
361
834
42
109
575
577
308
129
181
344
989
519
950
338
743
252
837
102
688
916
925
330
850
35
178
12
291
373
738
323
174
149
42
422
95
875
474
564
839
500
733
240
972
114
105
365
689
14
694
907
485
647
988
585
350
900
154
136
549
531
433
920
929
216
380
330
477
959
743
567
66
720
19
100
485
715
706
739
484
968
778
315
328
577
533
198
104
618
918
747
136
998
116
61
86
301
90
740
503
537
595
406
241
784
867
92
871
888
931
748
654
791
604
663
33
588
30
906
729
903
379
425
36
927
835
281
93
158
887
456
138
383
137
181
214
434
499
212
357
490
555
23
374
182
47
827
400
119
463
259
129
202
617
395
537
712
439
423
607
513
927
296
339
804
503
100
94
969
156
486
467
452
52
455
408
249
288
520
639
296
33
9
83
775
240
981
150
1001
615
262
963
688
530
816
862
668
199
49
96
753
164
41
265
586
119
598
874
959
344
376
533
150
123
345
767
772
933
346
82
947
890
477
129
831
76

891
105
870
697
675
678
914
753
214
300
196
204
15
125
469
698
508
201
885
562
873
832
849
550
942
235
878
262
756
35
271
15
804
510
353
497
355
13
872
585
226
362
728
684
859
582
113
780
564
348
828
246
899
69
225
860
321
237
840
382
919
386
274
186
825
595
921
980
526
320
958
530
204
501
735
684
735
96
739
201
642
921
767
637
395
334
713
94
720
378
868
121
741
527
734
1001
949
371
41
484
72
958
124
60
905
639
889
999
485
337
584
23
562
928
635
915
421
764
529
282
506
276
955
86
896
461
287
468
43
857
591
439
556
903
766
265
855
707
334
785
163
323
874
644
463
616
172
367
868
225
76
951
921
41
650
562
403
103
630
846
390
305
829
260
429
429
72
581
231
883
129
642
560
218
456
467
214
660
35
105
474
806
667
596
738
263
762
171
54
314
931
906
488
338
69
578
445
139
313
171
17
465
796
301
714
938
454
175
742
488
431
986
432
681
515
46
826
695
342
837
995
987
920
232
836
716
330
610
396
460
854
144
621
490
248
49
943
826
769
27
113
409
712
661
652
696
365
110
77
331
267
452
978
811
446
662

280
66
750
815
794
777
235
864
154
26
950
355
908
144
807
82
494
299
552
995
973
924
148
642
944
111
742
237
147
525
813
846
72
413
164
739
791
866
643
678
917
225
489
833
683
608
449
106
448
846
877
525
430
142
667
950
297
643
313
228
57
145
442
489
767
666
527
989
77
753
688
496
943
372
677
43
502
627
682
898
110
385
480
277
568
329
936
693
281
973
841
700
59
478
991
686
279
116
499
248
770
758
183
426
686
988
453
105
226
721
311
280
536
638
512
180
300
576
431
743
25
447
601
316
999
384
61
673
611
396
437
887
733
765
664
649
574
843
738
905
688
502
236
270
495
349
178
776
57
21
439
336
207
86
894
37
378
240
275
327
914
175
3
905
994
379
274
829
648
43
794
396
995
112
642
108
882
379
886
714
481
811
40
744
898
821
384
817
972
133
254
345
319
751
727
336
210
656
987
950
715
112
205
453
852
18
851
628
289
906
365
419
619
471
474
661
1001
790
362
150
511
208
545
894
698
418
419
343
36
165
551
224
455
775
48
1000
443
843
611
448
147
730
958
186
587
887
630
494
375
156
38
548
963
257
935


76
986
979
678
232
268
899
852
728
969
694
928
95
864
579
262
530
34
177
51
919
873
455
339
196
40
589
950
305
580
284
622
260
91
567
321
996
85
973
935
36
545
735
691
973
846
938
413
439
427
471
117
455
801
800
498
805
683
694
962
908
439
959
743
17
691
946
436
693
356
374
696
988
277
183
170
985
968
412
519
237
438
321
421
700
447
878
780
261
175
858
33
596
365
837
382
245
130
967
898
809
29
507
261
589
353
908
47
75
963
439
867
660
853
246
644
807
410
668
585
984
570
981
712
510
561
676
962
105
274
743
643
465
903
958
255
249
336
856
659
349
231
405
890
378
797
248
452
160
323
321
762
829
338
260
910
221
188
594
953
126
634
473
353
388
950
964
329
319
97
962
751
824
457
518
849
754
375
674
133
816
30
791
705
531
796
406
336
112
445
214
901
384
356
255
953
958
199
133
821
918
63
751
489
460
525
970
881
156
276
993
135
19
49
554
423
907
888
793
890
614
577
567
957
403
992
899
894
244
736
198
951
967
828
287
344
243
3
860
107
747
43
331
932
321
86
425
227
408
379
141
944
877
786
231
83

545
830
946
120
778
43
517
577
662
202
470
780
563
595
421
813
830
262
134
545
540
846
253
695
752
40
537
369
8
245
561
897
768
807
526
460
13
283
346
675
410
2
988
572
564
968
687
897
543
5
499
513
560
270
210
563
321
136
260
757
247
532
479
290
287
568
83
522
638
761
238
410
540
554
339
818
40
986
461
354
741
438
601
553
336
126
39
950
66
622
781
632
103
709
693
623
129
284
784
172
531
131
499
837
167
323
411
960
998
423
455
671
887
242
229
634
708
595
107
504
967
750
611
461
542
587
569
524
644
261
672
281
278
752
513
205
272
301
430
666
138
739
963
360
407
89
261
880
353
555
85
302
838
742
462
71
262
235
11
470
450
990
470
827
664
481
746
394
192
160
240
306
115
416
859
399
446
194
588
898
649
176
658
691
297
225
7
587
641
184
633
368
877
413
818
142
922
862
486
495
149
315
176
784
854
31
605
228
959
684
705
160
739
359
882
79
686
81
538
511
882
183
106
330
276
887
722
830
681
574
11
580
708
625
39
118
493
614
81
790
613
816
290
995
635
758
983
92
598
607
65
736
485
464
185
598
349

138
66
700
261
633
567
763
655
146
726
67
461
900
357
265
547
343
186
971
466
687
974
727
399
818
284
705
519
142
87
796
879
714
336
236
244
834
324
289
23
478
306
493
6
818
955
228
889
5
882
677
111
648
145
69
700
384
622
301
14
57
70
833
742
998
345
934
712
468
365
482
88
109
359
894
488
349
888
596
152
502
756
33
818
633
668
251
21
527
564
14
309
17
208
99
421
782
622
694
198
68
996
213
600
849
706
308
945
147
928
752
453
986
309
714
436
895
710
3
30
245
452
912
859
625
986
357
215
4
880
463
88
963
262
369
423
463
365
919
489
577
706
867
988
776
203
169
293
46
213
291
458
326
50
176
614
126
412
260
801
679
477
498
171
250
821
388
661
918
213
389
945
43
865
617
358
850
1001
53
769
159
123
565
581
347
537
877
495
470
192
973
647
19
6
304
946
408
99
955
627
331
673
758
374
396
172
792
210
152
328
455
261
686
829
509
619
829
189
906
107
947
977
913
909
878
633
219
609
207
703
294
541
989
102
708
80
386
739
934
115
352
474
98
970
1000
279
723
877
987
869
743
732
303
571
132
825
338
412
6

498
642
743
332
412
861
885
558
893
27
12
322
927
959
491
53
99
331
938
218
226
142
738
817
412
100
961
782
352
172
80
533
71
826
984
969
932
337
431
692
327
20
706
195
668
904
130
112
963
708
490
919
772
141
883
136
553
390
335
899
62
79
254
828
47
968
826
451
444
476
866
746
140
338
64
677
841
287
789
208
868
236
19
318
161
811
696
520
902
92
786
858
558
296
623
556
364
390
279
549
86
378
331
901
523
429
201
836
764
727
349
468
522
932
900
972
689
596
193
63
413
55
383
910
461
494
574
556
915
475
582
500
191
404
103
489
821
561
640
129
500
832
194
240
613
148
412
671
604
563
216
201
269
125
523
980
36
400
929
368
839
662
664
772
502
557
297
657
492
426
883
637
547
355
886
343
809
957
680
659
473
43
766
724
61
683
946
713
848
767
384
760
365
643
915
873
207
618
912
992
746
225
263
876
664
473
805
714
982
96
608
74
41
917
173
843
80
560
898
601
920
937
146
511
186
657
79
286
349
917
89
682
998
906
145
625
17
517
716
403
486
415
362
625
676
107
335
919
86
345
192
864
92
600
221
7
933
40

936
257
62
430
135
578
437
743
595
689
391
148
632
381
752
707
994
448
609
953
206
590
227
627
275
646
831
558
899
842
764
947
797
922
715
806
643
932
690
375
276
585
320
760
92
31
261
525
759
495
48
994
923
197
89
999
467
842
782
369
831
213
838
290
197
70
531
763
77
719
431
781
305
88
451
444
650
596
983
632
716
100
674
666
457
332
730
358
949
332
439
184
859
323
393
806
230
694
17
470
321
927
708
482
993
170
204
785
581
1000
931
305
579
809
359
137
679
830
242
398
291
510
206
331
315
47
33
122
587
700
567
885
510
4
82
604
891
507
850
108
205
263
139
360
949
504
29
461
744
479
499
491
86
459
933
223
490
968
73
939
783
175
895
830
124
636
972
599
542
125
801
143
712
812
225
680
607
967
129
787
341
983
414
402
416
445
920
33
442
834
109
391
185
627
384
753
443
603
410
46
651
329
465
141
361
540
982
760
8
229
307
457
176
256
865
257
751
104
139
410
505
77
739
689
76
358
874
602
359
244
29
874
141
328
851
325
711
979
368
505
16
442
528
570
334
895
683
780
850
498
907
275
322
422
670
347


421
733
548
120
815
382
914
303
750
332
137
174
722
498
972
188
365
596
143
623
909
479
507
669
915
62
651
346
12
4
603
458
956
601
717
300
192
192
225
96
60
286
681
885
651
811
733
127
2
531
740
295
794
618
662
77
405
902
226
345
469
926
51
500
425
786
29
681
492
204
7
265
226
406
346
636
523
487
384
771
139
778
289
692
24
936
822
217
234
331
137
17
614
171
501
35
688
332
338
781
828
107
533
594
189
399
523
578
812
751
361
895
156
247
226
536
86
239
909
431
135
862
376
172
229
519
543
675
757
88
118
958
863
248
626
233
923
892
647
198
373
925
170
947
991
320
967
709
490
475
605
200
870
231
187
961
443
456
141
227
944
660
365
261
447
622
700
371
820
213
741
441
783
270
70
354
425
315
691
94
550
678
289
250
357
614
320
126
345
932
846
521
313
466
69
284
934
154
951
364
28
957
140
483
376
763
542
677
114
372
797
310
995
833
841
84
383
377
280
363
403
339
993
626
661
367
313
450
887
338
480
26
929
66
779
882
252
927
898
846
931
537
74
265
452
510
815
235
278
243
389
379
11
901
190
568
470

626
794
964
456
502
473
792
769
856
835
329
635
679
542
852
852
143
212
706
329
882
881
714
751
10
168
761
243
949
49
583
646
190
509
766
863
646
126
28
669
25
456
33
285
393
734
995
686
154
261
916
308
56
774
142
310
519
31
101
898
574
168
426
763
337
614
250
248
318
664
485
370
12
503
279
717
150
26
66
629
10
662
162
795
301
254
352
21
502
587
869
832
326
823
585
692
856
622
263
329
139
475
713
210
267
729
700
519
270
609
695
331
957
985
932
60
910
416
975
840
374
754
500
458
836
577
677
320
169
166
280
745
196
733
287
963
615
211
34
675
425
530
578
250
249
295
71
197
617
452
191
809
422
144
439
213
739
256
140
769
955
150
320
27
676
66
477
989
254
674
332
708
815
62
86
691
301
468
690
250
400
229
339
615
667
771
914
24
842
876
559
702
520
262
525
787
491
419
703
378
450
494
739
860
857
441
896
697
765
972
339
673
612
861
82
795
540
927
736
112
967
225
406
145
432
648
145
563
347
765
827
878
557
516
997
965
35
730
455
739
154
378
849
53
597
542
100
994
206
203
324
282
737
71
741
240


52
613
269
568
466
37
600
602
803
722
80
180
68
841
222
724
398
171
263
475
723
96
913
552
45
411
675
433
174
184
745
205
750
59
573
816
278
143
467
148
506
303
143
529
347
311
242
957
426
805
556
150
596
466
165
877
406
593
112
981
724
299
234
279
424
6
686
281
790
549
163
545
466
771
693
992
818
227
553
245
32
243
950
183
106
683
974
482
965
322
500
312
358
465
110
297
689
899
768
337
266
903
662
916
866
968
378
236
869
445
331
868
201
507
821
764
177
678
690
793
635
229
539
658
485
310
398
859
260
852
78
465
557
551
255
292
11
756
872
518
135
500
819
340
209
889
45
249
821
317
249
398
645
383
986
678
781
161
713
300
382
645
311
60
778
953
262
241
291
242
751
585
759
158
654
895
902
245
67
903
550
243
810
364
695
505
319
818
473
16
763
750
169
424
745
248
33
268
66
926
186
895
835
873
699
257
210
195
357
748
504
980
999
269
106
412
862
701
7
796
344
380
933
826
611
113
384
987
949
149
425
632
607
880
498
816
791
576
580
983
773
805
602
946
641
426
475
750
985
409
553
699
137
886
462


168
403
834
687
562
332
153
279
770
207
380
352
825
375
372
536
299
789
599
72
216
574
56
204
577
426
149
377
278
858
867
166
785
253
39
181
798
665
38
554
511
971
699
42
711
863
834
630
117
765
378
543
559
572
284
715
634
162
159
147
525
126
989
28
174
977
314
63
884
189
893
706
671
936
284
93
635
841
456
760
451
789
767
860
429
350
396
517
791
635
789
521
404
973
902
53
643
59
263
607
747
217
973
677
418
718
180
756
586
434
153
548
334
232
529
526
155
387
302
188
822
795
500
333
908
950
240
855
271
888
713
75
404
800
636
29
578
34
769
756
959
849
7
529
849
667
254
67
911
36
41
557
993
601
163
626
705
22
871
103
812
122
9
384
531
146
913
572
839
474
971
442
44
606
274
494
698
831
592
277
309
137
44
879
946
609
75
659
158
967
637
417
846
892
53
774
531
381
503
664
232
562
764
685
69
147
489
965
466
596
964
821
488
399
448
17
522
21
130
89
821
434
949
868
592
420
559
194
469
339
648
754
484
467
91
21
222
347
106
546
703
553
477
858
103
227
132
110
825
73
275
319
839
685
604
486
328
859


843
535
143
376
844
389
909
578
785
708
805
271
847
557
191
499
730
33
706
208
560
202
594
137
201
574
393
384
685
469
97
170
698
56
459
685
371
999
291
119
473
251
344
194
598
540
935
929
951
178
690
682
625
398
789
82
557
886
203
80
311
244
85
329
745
9
563
753
451
380
273
86
916
729
211
990
720
622
287
95
695
172
553
762
586
183
123
649
14
815
224
935
179
174
351
217
981
515
83
191
106
260
177
217
839
827
974
358
220
859
126
985
681
716
8
182
339
309
810
490
647
959
861
406
870
485
14
125
707
749
790
908
919
874
311
342
5
324
24
917
110
443
829
733
450
580
824
852
29
149
704
599
22
119
470
98
266
921
880
180
370
734
213
234
622
703
403
771
514
162
354
769
286
731
99
759
479
554
799
648
105
66
24
672
663
625
405
37
736
974
371
103
972
356
345
854
577
852
631
739
77
677
98
985
221
674
996
640
529
879
75
355
712
890
261
90
585
764
662
273
216
345
941
228
750
226
180
845
44
875
352
11
976
377
671
942
575
601
167
623
520
186
815
899
838
673
678
688
263
26
353
907
197
987
479
101
426
26
5

81
306
972
221
589
53
384
452
713
535
162
846
840
821
86
655
820
639
110
568
820
593
17
802
598
788
990
571
305
465
698
762
779
954
895
626
123
491
603
177
621
108
360
840
10
707
356
12
655
670
616
884
649
846
964
663
164
404
759
434
666
984
332
143
294
385
378
134
536
285
402
680
454
532
279
920
789
610
470
84
924
267
804
21
737
977
462
751
867
507
51
222
64
495
876
94
644
639
483
820
328
531
684
251
805
28
218
529
22
853
601
617
839
978
482
28
68
262
233
341
980
622
727
37
728
408
140
891
787
800
478
825
151
113
870
990
797
786
188
364
545
174
564
798
636
628
865
968
39
832
660
871
878
594
824
37
743
520
758
583
147
386
210
719
374
341
997
582
706
306
44
955
400
186
387
595
563
360
738
991
950
191
840
31
412
687
191
49
94
977
515
655
17
619
906
913
543
719
578
234
985
172
319
155
101
837
600
816
142
489
624
731
990
907
788
356
353
232
938
814
427
28
591
72
439
416
471
908
427
372
867
763
953
210
784
3
127
138
331
11
412
835
528
97
424
547
15
344
464
355
508
598
246
563
152
315
691
50

900
727
997
184
497
901
819
832
314
538
262
844
790
339
778
366
623
831
602
746
2
919
515
890
829
716
63
792
913
261
654
561
823
230
197
105
98
706
329
525
856
199
830
414
242
857
916
572
821
665
934
935
885
649
332
52
819
549
133
16
69
719
729
1001
238
180
543
430
242
724
697
576
133
663
105
836
782
91
370
5
104
391
339
523
497
299
425
632
878
354
18
480
646
573
231
474
277
276
339
922
946
359
278
631
236
137
849
922
549
891
53
528
160
978
90
712
152
19
344
367
933
103
337
801
135
625
667
286
431
893
944
500
377
793
341
348
714
839
901
815
425
255
256
373
3
604
300
465
94
988
532
719
323
736
686
693
357
249
234
297
590
737
129
740
402
878
569
954
122
975
116
390
140
364
521
323
702
933
826
121
594
790
582
614
655
563
144
592
551
553
131
811
980
593
490
257
792
759
795
623
198
748
286
394
540
732
600
400
664
20
808
205
447
193
811
383
624
747
747
307
868
249
288
462
677
480
267
453
185
657
93
46
870
605
870
768
111
63
77
393
125
676
768
444
946
238
797
747
853
544
596
697
758
923
266
3

34
801
387
452
663
685
183
20
482
32
381
119
631
520
174
693
35
629
456
462
752
776
47
636
531
765
248
421
188
863
347
442
174
659
354
354
178
758
869
967
51
357
54
962
514
76
107
881
834
364
741
121
203
375
283
975
204
424
525
760
356
134
454
332
598
666
883
403
686
821
669
964
558
873
217
233
984
456
118
807
425
666
576
935
520
444
337
204
695
223
328
772
615
345
606
206
486
130
900
311
300
735
161
717
309
488
504
558
318
532
987
162
509
577
16
700
337
61
534
465
25
660
795
842
220
254
270
31
311
875
770
290
937
221
97
538
25
155
521
617
805
945
318
155
533
583
719
718
8
9
447
879
176
331
54
583
839
728
330
860
159
124
129
227
579
924
312
564
516
76
348
145
657
865
705
678
913
625
801
389
315
763
562
443
391
211
450
772
531
711
79
36
140
591
801
897
627
999
199
134
166
586
734
602
80
96
455
277
591
737
978
619
616
916
22
646
666
547
323
313
366
913
8
267
759
552
412
22
686
517
713
922
355
640
720
523
98
576
907
758
552
35
582
39
112
525
455
515
228
817
891
674
937
204
349
353
987
446

548
919
9
275
198
543
604
912
829
51
745
801
592
259
732
824
591
341
913
461
364
917
839
18
573
660
192
173
546
551
516
148
501
74
375
759
425
349
856
800
803
989
635
738
727
200
809
902
725
947
791
503
753
84
4
924
664
863
881
638
286
804
64
571
235
885
881
591
958
247
212
865
287
407
347
382
291
25
817
58
274
286
31
955
218
969
600
292
225
211
843
417
180
649
899
891
18
67
688
754
448
242
458
244
678
971
689
606
473
955
712
964
85
625
797
276
473
788
305
261
151
242
69
997
913
292
498
33
31
599
174
11
638
310
522
874
118
250
263
974
457
520
653
827
69
724
324
559
874
247
21
572
812
862
577
434
364
717
336
550
25
452
3
987
598
580
813
468
506
917
604
157
140
790
113
832
101
108
944
925
951
329
105
923
822
341
938
26
482
569
839
479
535
887
145
20
260
208
619
927
802
863
628
542
24
274
630
385
821
159
452
814
16
605
154
152
739
604
842
525
674
106
253
510
943
551
609
895
171
652
308
299
692
984
725
43
600
843
956
62
491
178
768
496
399
586
158
673
659
926
99
629
655
554
862
32
98
934
8

476
346
763
953
8
810
38
466
956
410
697
259
556
939
720
427
360
165
451
388
36
502
947
804
424
696
565
751
154
529
710
402
592
726
404
796
849
143
370
934
571
12
819
785
688
436
294
886
347
993
207
952
853
779
254
592
550
765
506
590
445
491
715
271
233
384
6
707
835
86
165
145
278
615
674
4
158
982
215
923
899
740
45
963
202
904
783
542
212
435
93
777
258
588
697
186
766
898
423
233
658
284
473
605
100
26
481
94
498
289
237
927
191
970
752
464
932
965
988
283
975
170
513
904
137
233
308
548
123
555
639
762
464
625
511
764
910
281
258
446
296
155
391
508
805
579
316
213
955
75
175
802
434
411
836
394
178
495
214
670
266
423
131
672
801
761
916
835
901
178
281
481
703
332
92
947
53
437
174
554
281
156
132
877
668
305
855
983
230
998
922
358
439
301
949
70
70
957
758
483
576
921
816
75
59
530
282
628
702
556
474
299
87
680
403
188
843
75
629
872
463
687
879
433
33
542
283
447
820
800
408
938
586
148
271
629
593
748
896
611
923
677
368
842
338
571
668
768
720
656
984
46
273
332
323
663
2

377
608
794
912
124
918
527
690
813
6
618
473
447
763
8
603
174
442
401
709
173
99
595
194
302
336
113
538
695
215
4
365
401
232
428
631
767
970
403
38
162
698
703
139
674
122
608
228
420
683
394
179
375
585
195
341
991
67
370
158
327
171
874
482
942
776
834
855
639
676
571
125
614
52
330
531
286
639
538
583
191
72
707
463
745
558
664
8
660
925
851
993
606
414
690
164
177
208
94
347
938
841
763
280
944
435
269
528
114
17
857
512
300
790
1000
480
831
231
479
236
246
852
294
196
741
863
850
178
984
50
462
380
702
965
576
109
297
666
294
922
985
504
627
977
809
453
669
865
903
676
813
310
752
156
96
898
360
985
978
90
768
102
376
267
533
137
133
504
11
397
606
952
130
864
839
925
555
795
360
548
359
8
810
492
452
275
634
110
606
319
971
289
221
594
427
700
530
630
654
561
825
1000
895
585
410
900
390
794
212
870
588
790
113
220
863
607
685
351
675
202
499
169
907
356
216
567
36
788
389
689
631
432
718
517
479
758
539
790
869
310
245
108
994
494
617
391
198
563
487
256
666
8
639
950
628
64

447
381
774
72
611
866
407
248
974
943
981
810
430
273
343
496
688
283
135
605
248
730
799
239
451
511
131
862
522
996
558
448
332
988
511
743
786
810
924
13
699
138
484
240
449
672
5
457
849
521
798
313
874
538
487
481
954
442
189
357
317
230
786
653
991
945
431
213
976
17
519
247
358
325
70
475
396
200
741
792
937
244
962
148
928
273
324
957
817
761
585
460
346
201
993
962
370
833
657
838
314
100
398
265
197
454
125
421
602
853
108
129
351
694
755
117
618
402
380
41
983
693
217
305
97
550
863
609
143
907
373
230
876
993
171
738
39
393
21
49
878
853
683
123
973
498
748
140
130
816
331
96
499
638
44
300
353
681
470
212
493
13
217
720
843
288
891
76
384
164
373
187
122
320
813
672
881
153
969
68
872
484
28
480
632
661
670
40
85
445
851
998
438
169
608
576
680
54
488
463
203
180
911
554
103
722
327
161
756
933
597
687
11
457
137
352
807
831
695
875
317
972
796
946
284
591
535
194
649
967
487
344
696
69
4
774
14
651
648
291
658
960
691
649
356
191
503
667
745
378
106
200
232
288
962
604
2

In [10]:
%%timeit -n 10
s=pd.Series(np.random.randint(0,1000,10000))
s+=2

The slowest run took 17.44 times longer than the fastest. This could mean that an intermediate result is being cached.
1.79 ms ± 2.94 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


One last note on using the indexing operators to access series data. **The .loc attribute lets you not only modify data in place, but also add new data as well.** If the value you pass in as the index doesn't exist, then a new entry is added. And keep in mind, indices can have mixed types. While it's important to be aware of the typing going on underneath, Pandas will automatically change the underlying NumPy types as appropriate.

In [24]:
s=[1,2,3]
ss=pd.Series(s)
ss.loc['Animal']='Bears'
ss

0             1
1             2
2             3
Animal    Bears
dtype: object

Up until now I've shown only examples of a series where the index values were unique. I want to end this lecture by showing an example where index values are not unique, and this makes data frames different, conceptually, that a relational database might be.

In [26]:
original_sports= pd.Series({'Archery':'Bhutan','Golf':'Scotland','Sumo':'Japan'
                           ,'Shooting':'Italy',})

cricket_loving_countries=pd.Series(['Australia','India','Pakistan','England'],index=['Cricket'
                                    ,'Cricket','Cricket','Cricket'])

all_countries=original_sports.append(cricket_loving_countries)

In [27]:
original_sports

Archery       Bhutan
Golf        Scotland
Sumo           Japan
Shooting       Italy
dtype: object

In [28]:
cricket_loving_countries

Cricket    Australia
Cricket        India
Cricket     Pakistan
Cricket      England
dtype: object

In [29]:
all_countries

Archery        Bhutan
Golf         Scotland
Sumo            Japan
Shooting        Italy
Cricket     Australia
Cricket         India
Cricket      Pakistan
Cricket       England
dtype: object

## The DataFrame Data Structure

The DataFrame data structure is the heart of the Panda's library. It's a primary object that you'll be working with in data analysis and cleaning tasks.  
  
  **The DataFrame is conceptually a two-dimensional series object, where there's an index and multiple columns of content, with each column having a label.** In fact, the distinction between a column and a row is really only a conceptual distinction. And you can think of the DataFrame itself as simply a two-axes labeled array.

In [12]:
purchase_1 = pd.Series({'Name': 'Chris',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Kevyn',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Vinod',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})
df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])
df.head()

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

Similar to the series, we can extract data using the iLock and Lock attributes. Because the DataFrame is two-dimensional, passing a single value to the lock indexing operator will return series if there's only one row to return

In [31]:
df.loc['Store 2']

Name                  Vinod
Item Purchased    Bird Seed
Cost                      5
Name: Store 2, dtype: object

It's important to remember that the indices and column names along either axes, horizontal or vertical, could be non-unique. For instance, in this example, we see two purchase records for Store 1 as different rows. If we use a single value with the DataFrame lock attribute, multiple rows of the DataFrame will return, **not as a new series, but as a new DataFrame.**

In [34]:
df.loc['Store 1']

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5

One of the powers of the Panda's DataFrame is that you can quickly **select data based on multiple axes.**

In [14]:
df.loc['Store 1','Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

**What if we just wanted to do column selection and just get a list of all of the costs?**

In [39]:
df.T

Store 1       Store 1    Store 2
Name               Chris         Kevyn      Vinod
Item Purchased  Dog Food  Kitty Litter  Bird Seed
Cost                22.5           2.5          5

In [41]:
df.T.loc['Cost']

Store 1    22.5
Store 1     2.5
Store 2       5
Name: Cost, dtype: object

Here's another method. As we saw, **.loc does row selection, and it can take two parameters, the row index and the list of column names. .loc also supports slicing.** If we wanted to select all rows, we can use a column to indicate a full slice from beginning to end. And then add the column name as the second parameter as a string. In fact, if we wanted to include multiply columns, we could do so in a list. And Pandas will bring back only the columns we have asked for.

In [15]:
df.loc[:,['Name','Cost']]
# df[['Name','Cost']]

Name  Cost
Store 1  Chris  22.5
Store 1  Kevyn   2.5
Store 2  Vinod   5.0

It's easy to delete data in series and DataFrames, and we can use the **drop function** to do so. This function **takes a single parameter, which is the index or roll label, to drop.**  
  
  The drop function doesn't change the DataFrame by default. And instead, returns to you a copy of the DataFrame with the given rows removed. We can see that our original DataFrame is still intact.

In [20]:
print(df.drop('Store 1'))
df

          Name Item Purchased  Cost
Store 2  Vinod      Bird Seed   5.0


Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

In [19]:
print(df.drop('Name',axis=1))
df

        Item Purchased  Cost
Store 1       Dog Food  22.5
Store 1   Kitty Litter   2.5
Store 2      Bird Seed   5.0


Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

Drop has two interesting optional parameters. The first is called **in place**, and if it's set to true, the DataFrame will be updated in place, instead of a copy being returned. The second parameter is the **axes**, which should be dropped. By default, this **value is 0, indicating the row axes**. But you could change it to 1 if you want to drop a column.

There is a second way to **drop a column**, however. And that's directly through the use of the indexing operator, using the **del keyword.**

In [17]:
copy_df=df.copy()
del copy_df['Name']
copy_df

Item Purchased  Cost
Store 1       Dog Food  22.5
Store 1   Kitty Litter   2.5
Store 2      Bird Seed   5.0

Finally, adding a new column to the DataFrame is as easy as assigning it to some value.

In [59]:
df['Location']=None
df

Name Item Purchased  Cost Location
Store 1  Chris       Dog Food  22.5     None
Store 1  Kevyn   Kitty Litter   2.5     None
Store 2  Vinod      Bird Seed   5.0     None

**Question:1 For the purchase records from the pet store, how would you update the DataFrame, applying a discount of 20% across all the values in the 'Cost' column?**

In [60]:
df['Cost'] *= 0.8
df

Name Item Purchased  Cost Location
Store 1  Chris       Dog Food  18.0     None
Store 1  Kevyn   Kitty Litter   2.0     None
Store 2  Vinod      Bird Seed   4.0     None

## DataFrame Indexing and Loading

**The common work flow is to read your data into a DataFrame then reduce this DataFrame to the particular columns or rows that you're interested in working with.** As you've seen, the Panda's toolkit tries to give you views on a DataFrame. This is much faster than copying data and much more memory efficient too.  
  
  But it does mean that if you're **manipulating the data you have to be aware that any changes to the DataFrame you're working on may have an impact on the base data frame you used originally.**

Here's an example using our same purchasing DataFrame from earlier. We can create a series based on just the cost category using the square brackets. Then we can increase the cost in this series using broadcasting. Now if we look at our original DataFrame, we see those costs have risen as well. This is an important consideration to watch out for. If you want to explicitly use a copy, then you should consider calling the copy method on the DataFrame for it first.

In [21]:
costs = df['Cost']
costs

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

In [22]:
costs+=2
costs

Store 1    24.5
Store 1     4.5
Store 2     7.0
Name: Cost, dtype: float64

In [23]:
df

Name Item Purchased  Cost
Store 1  Chris       Dog Food  24.5
Store 1  Kevyn   Kitty Litter   4.5
Store 2  Vinod      Bird Seed   7.0

What happens here is that when the Jupyter notebook sees a line beginning with an exclamation mark, it sends the rest of the line to the operating system shell for evaluation.

In [24]:
!cat olympics.csv

0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
,№ Summer,01 !,02 !,03 !,Total,№ Winter,01 !,02 !,03 !,Total,№ Games,01 !,02 !,03 !,Combined total
Afghanistan (AFG),13,0,0,2,2,0,0,0,0,0,13,0,0,2,2
Algeria (ALG),12,5,2,8,15,3,0,0,0,0,15,5,2,8,15
Argentina (ARG),23,18,24,28,70,18,0,0,0,0,41,18,24,28,70
Armenia (ARM),5,1,2,9,12,6,0,0,0,0,11,1,2,9,12
Australasia (ANZ) [ANZ],2,3,4,5,12,0,0,0,0,0,2,3,4,5,12
Australia (AUS) [AUS] [Z],25,139,152,177,468,18,5,3,4,12,43,144,155,181,480
Austria (AUT),26,18,33,35,86,22,59,78,81,218,48,77,111,116,304
Azerbaijan (AZE),5,6,5,15,26,5,0,0,0,0,10,6,5,15,26
Bahamas (BAH),15,5,2,5,12,0,0,0,0,0,15,5,2,5,12
Bahrain (BRN),8,0,0,1,1,0,0,0,0,0,8,0,0,1,1
Barbados (BAR) [BAR],11,0,0,1,1,0,0,0,0,0,11,0,0,1,1
Belarus (BLR),5,12,24,39,75,6,6,4,5,15,11,18,28,44,90
Belgium (BEL),25,37,52,53,142,20,1,1,3,5,45,38,53,56,147
Bermuda (BER),17,0,0,1,1,7,0,0,0,0,24,0,0,1,1
Bohemia (BOH) [BOH] [Z],3,0,1,3,4,0,0,0,0,0,3,0,1,3,4
Botswana (BOT),9,0,1,0,1,0,0,0,0,0,9,0,1,0,1
Brazil (BRA),

We can read this into a DataFrame by calling the read_csv function of the module. When we look at the DataFrame we see that the first cell has an NaN in it since it's an empty value, and the rows have been automatically indexed for us.

In [6]:
df = pd.read_csv('olympics.csv')
df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  № Summer  01 !  02 !  03 !  Total  № Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  № Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

It seems pretty clear that the first row of data in the DataFrame is what we really want to see as the column names. It also seems like the first column in the data is the country name, which we would like to make an index.  
Read csv has a number of parameters that we can use to indicate to Pandas how rows and columns should be labeled.  
  
  For instance, we can use the index col to indicate which column should be the index and we can also use the header parameter to indicate which row from the data file should be used as the header.

In [17]:
df=pd.read_csv('olympics.csv',index_col=0,skiprows=1)
df

№ Summer  01 !  02 !  03 !  \
Afghanistan (AFG)                                   13     0     0     2   
Algeria (ALG)                                       12     5     2     8   
Argentina (ARG)                                     23    18    24    28   
Armenia (ARM)                                        5     1     2     9   
Australasia (ANZ) [ANZ]                              2     3     4     5   
...                                                ...   ...   ...   ...   
Independent Olympic Participants (IOP) [IOP]         1     0     1     2   
Zambia (ZAM) [ZAM]                                  12     0     1     1   
Zimbabwe (ZIM) [ZIM]                                12     3     4     1   
Mixed team (ZZX) [ZZX]                               3     8     5     4   
Totals                                              27  4809  4775  5130   

                                              Total  № Winter  01 !.1  02 !.1  \
Afghanistan (AFG)                                 2         0       0       0   
Algeria (ALG)                                    15         3       0       0   
Argentina (ARG)                                  70        18       0       0   
Armenia (ARM)                                    12         6       0       0   
Australasia (ANZ) [ANZ]                          12         0       0       0   
...                                             ...       ...     ...     ...   
Independent Olympic Participants (IOP) [IOP]      3         0       0       0   
Zambia (ZAM) [ZAM]                                2         0       0       0   
Zimbabwe (ZIM) [ZIM]                              8         1       0       0   
Mixed team (ZZX) [ZZX]                           17         0       0       0   
Totals                                        14714        22     959     958   

                                              03 !.1  Total.1  № Games  \
Afghanistan (AFG)                                  0        0       13   
Algeria (ALG)                                      0        0       15   
Argentina (ARG)                                    0        0       41   
Armenia (ARM)                                      0        0       11   
Australasia (ANZ) [ANZ]                            0        0        2   
...                                              ...      ...      ...   
Independent Olympic Participants (IOP) [IOP]       0        0        1   
Zambia (ZAM) [ZAM]                                 0        0       12   
Zimbabwe (ZIM) [ZIM]                               0        0       13   
Mixed team (ZZX) [ZZX]                             0        0        3   
Totals                                           948     2865       49   

                                              01 !.2  02 !.2  03 !.2  \
Afghanistan (AFG)                                  0       0       2   
Algeria (ALG)                                      5       2       8   
Argentina (ARG)                                   18      24      28   
Armenia (ARM)                                      1       2       9   
Australasia (ANZ) [ANZ]                            3       4       5   
...                                              ...     ...     ...   
Independent Olympic Participants (IOP) [IOP]       0       1       2   
Zambia (ZAM) [ZAM]                                 0       1       1   
Zimbabwe (ZIM) [ZIM]                               3       4       1   
Mixed team (ZZX) [ZZX]                             8       5       4   
Totals                                          5768    5733    6078   

                                              Combined total  
Afghanistan (AFG)                                          2  
Algeria (ALG)                                             15  
Argentina (ARG)                                           70  
Armenia (ARM)                                             12  
Australasia (ANZ) [ANZ]                                   12  
...                         

 If we head to the page we could see that instead of running gold, silver and bronze in the pages, these nice little icons with a one, a two, and a three in them In our csv file these were represented with the strings 01 !, 02 !, and so on.  
   
   But this labeling isn't really as clear as it could be, so we should clean up the data file. We can of course do this just by going and editing the CSV file directly, but we can also set the column names using the Pandas name property.

In [20]:
df.columns

Index(['№ Summer', '01 !', '02 !', '03 !', 'Total', '№ Winter', '01 !.1',
       '02 !.1', '03 !.1', 'Total.1', '№ Games', '01 !.2', '02 !.2', '03 !.2',
       'Combined total'],
      dtype='object')

In [4]:
for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold' + col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver' + col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze' + col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#' + col[1:]}, inplace=True) 

df()

TypeError: 'DataFrame' object is not callable

In [28]:
(df['Gold'] - df['Gold.1']).idxmax()

'Totals'

## Querying a DataFrame

Before we talk about how to query data frames, we need to talk about Boolean masking.**Boolean masking is the heart of fast and efficient querying in NumPy. It's analogous a bit to masking used in other computational areas.**
  
  
A Boolean mask is an array which can be of one dimension like a series, or two dimensions like a data frame, where each of the values in the array are either true or false. This array is essentially overlaid on top of the data structure that we're querying. And any cell aligned with the true value will be admitted into our final result, and any sign aligned with a false value will not.

Boolean masks are created by applying operators directly to the pandas series or DataFrame objects. For instance, in our Olympics data set, you might be interested in seeing only those countries who have achieved a gold medal at the summer Olympics. To build a Boolean mask for this query, we project the gold column using the indexing operator and apply the greater than operator with a comparison value of zero. This is essentially broadcasting a comparison operator, greater than, with the results being returned as a Boolean series. The resultant series is indexed where the value of each cell is either true or false depending on whether a country has won at least one gold medal, and the index is the country name.

In [39]:
#The below is boolean mask for Gold column
df['Gold']>0

Afghanistan (AFG)                               False
Algeria (ALG)                                    True
Argentina (ARG)                                  True
Armenia (ARM)                                    True
Australasia (ANZ) [ANZ]                          True
                                                ...  
Independent Olympic Participants (IOP) [IOP]    False
Zambia (ZAM) [ZAM]                              False
Zimbabwe (ZIM) [ZIM]                             True
Mixed team (ZZX) [ZZX]                           True
Totals                                           True
Name: Gold, Length: 147, dtype: bool

What we want to do next is overlay that mask on the data frame. We can do this using the where function. The where function takes a Boolean mask as a condition, applies it to the data frame or series, and returns a new data frame or series of the same shape. Let's apply this Boolean mask to our Olympics data and create a data frame of only those countries who have won a gold at a summer games.

In [40]:
#Creating a seperate df that matches our boolean mask
only_gold=df.where(df['Gold']>0)
only_gold.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)             NaN   NaN     NaN     NaN    NaN       NaN   
Algeria (ALG)                12.0   5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)              23.0  18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                 5.0   1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]       2.0   3.0     4.0     5.0   12.0       0.0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)           NaN       NaN       NaN      NaN      NaN     NaN   
Algeria (ALG)               0.0       0.0       0.0      0.0     15.0     5.0   
Argentina (ARG)             0.0       0.0       0.0      0.0     41.0    18.0   
Armenia (ARM)               0.0       0.0       0.0      0.0     11.0     1.0   
Australasia (ANZ) [ANZ]     0.0       0.0       0.0      0.0      2.0     3.0   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)             NaN       NaN             NaN  
Algeria (ALG)                 2.0       8.0            15.0  
Argentina (ARG)              24.0      28.0            70.0  
Armenia (ARM)                 2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       4.0       5.0            12.0

We see that the resulting data frame keeps the original indexed values, and only data from countries that met the condition are retained. All of the countries which did not meet the condition have NaN data instead. This is okay. Most statistical functions built into the data frame object ignore values of NaN.

For instance, if we call the df.count on the only gold data frame, we see that there are 100 countries which have had gold medals awarded at the summer games, while if we call count on the original data frame, we see that there are 147 countries total.

In [41]:
only_gold['Gold'].count()

100

In [42]:
df['Gold'].count()

147

Often we want to drop those rows which have no data. To do this, we can use the drop NA function. You can optionally provide drop NA the axes it should be considering. Remember that the axes is just an indicator for the columns or rows and that the default is zero, which means rows.

In [43]:
only_gold.dropna()

# Summer    Gold  Silver  Bronze    Total  \
Algeria (ALG)                  12.0     5.0     2.0     8.0     15.0   
Argentina (ARG)                23.0    18.0    24.0    28.0     70.0   
Armenia (ARM)                   5.0     1.0     2.0     9.0     12.0   
Australasia (ANZ) [ANZ]         2.0     3.0     4.0     5.0     12.0   
Australia (AUS) [AUS] [Z]      25.0   139.0   152.0   177.0    468.0   
...                             ...     ...     ...     ...      ...   
Venezuela (VEN)                17.0     2.0     2.0     8.0     12.0   
Yugoslavia (YUG) [YUG]         16.0    26.0    29.0    28.0     83.0   
Zimbabwe (ZIM) [ZIM]           12.0     3.0     4.0     1.0      8.0   
Mixed team (ZZX) [ZZX]          3.0     8.0     5.0     4.0     17.0   
Totals                         27.0  4809.0  4775.0  5130.0  14714.0   

                           # Winter  Gold.1  Silver.1  Bronze.1  Total.1  \
Algeria (ALG)                   3.0     0.0       0.0       0.0      0.0   
Argentina (ARG)                18.0     0.0       0.0       0.0      0.0   
Armenia (ARM)                   6.0     0.0       0.0       0.0      0.0   
Australasia (ANZ) [ANZ]         0.0     0.0       0.0       0.0      0.0   
Australia (AUS) [AUS] [Z]      18.0     5.0       3.0       4.0     12.0   
...                             ...     ...       ...       ...      ...   
Venezuela (VEN)                 4.0     0.0       0.0       0.0      0.0   
Yugoslavia (YUG) [YUG]         14.0     0.0       3.0       1.0      4.0   
Zimbabwe (ZIM) [ZIM]            1.0     0.0       0.0       0.0      0.0   
Mixed team (ZZX) [ZZX]          0.0     0.0       0.0       0.0      0.0   
Totals                         22.0   959.0     958.0     948.0   2865.0   

                           # Games  Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                 15.0     5.0       2.0       8.0            15.0  
Argentina (ARG)               41.0    18.0      24.0      28.0            70.0  
Armenia (ARM)                 11.0     1.0       2.0       9.0            12.0  
Australasia (ANZ) [ANZ]        2.0     3.0       4.0       5.0            12.0  
Australia (AUS) [AUS] [Z]     43.0   144.0     155.0     181.0           480.0  
...                            ...     ...       ...       ...             ...  
Venezuela (VEN)               21.0     2.0       2.0       8.0            12.0  
Yugoslavia (YUG) [YUG]        30.0    26.0      32.0      29.0            87.0  
Zimbabwe (ZIM) [ZIM]          13.0     3.0       4.0       1.0             8.0  
Mixed team (ZZX) [ZZX]         3.0     8.0       5.0       4.0            17.0  
Totals                        49.0  5768.0    5733.0    6078.0         17579.0  

[100 rows x 15 columns]

One more thing to keep in mind if you're not used to Boolean or bit masking for data reduction. The output of two Boolean masks being compared with logical operators is another Boolean mask. This means that you can chain together a bunch of and/or statements in order to create more complex queries, and the result is a single Boolean mask.

In [44]:
#No of Countries that have won Gold in Summmeror Winter olympics
len(df[(df['Gold']>0) | (df['Gold.1'])]>0)

101

In [45]:
#Countries that have won Gold only in Winter Olympics
df[(df['Gold.1']>0) & (df['Gold']==0)]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Liechtenstein (LIE)        16     0       0       0      0        18       2   

                     Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
Liechtenstein (LIE)         2         5        9       34       2         2   

                     Bronze.2  Combined total  
Liechtenstein (LIE)         5               9

## Indexing DataFrames

Both series and DataFrames can have indices applied to them. The index is essentially a row level label, and we know that rows correspond to axis zero.  

 Indices can either be inferred, such as when we create a new series without an index, in which case we get numeric values, or they can be set explicitly, like when we use the dictionary object to create the series, or when we loaded data from the CSV file and specified the header.  
   
 **Another option for setting an index is to use the set_index function.** This function takes a list of columns and promotes those columns to an index. Set index is a destructive process, it doesn't keep the current index. If you want to keep the current index, you need to manually create a new column and copy into it values from the index attribute.

In [46]:
#Making the current index 'Countries' as one of the usual column
df['country']=df.index

#setting 'Gold' column as new index of the df
df=df.set_index('Gold')
df.head()

# Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  Bronze.1  \
Gold                                                                          
0           13       0       2      2         0       0         0         0   
5           12       2       8     15         3       0         0         0   
18          23      24      28     70        18       0         0         0   
1            5       2       9     12         6       0         0         0   
3            2       4       5     12         0       0         0         0   

      Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
Gold                                                                 
0           0       13       0         0         2               2   
5           0       15       5         2         8              15   
18          0       41      18        24        28              70   
1           0       11       1         2         9              12   
3           0        2       3         4         5              12   

                      country  
Gold                           
0           Afghanistan (AFG)  
5               Algeria (ALG)  
18            Argentina (ARG)  
1               Armenia (ARM)  
3     Australasia (ANZ) [ANZ]

We can get rid of the index completely by calling the function **reset_index.** This promotes the index into a column and creates a default numbered index.

In [47]:
#Getting rid of the current index
df=df.reset_index()
df.head()

Gold  # Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
0     0        13       0       2      2         0       0         0   
1     5        12       2       8     15         3       0         0   
2    18        23      24      28     70        18       0         0   
3     1         5       2       9     12         6       0         0   
4     3         2       4       5     12         0       0         0   

   Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
0         0        0       13       0         0         2               2   
1         0        0       15       5         2         8              15   
2         0        0       41      18        24        28              70   
3         0        0       11       1         2         9              12   
4         0        0        2       3         4         5              12   

                   country  
0        Afghanistan (AFG)  
1            Algeria (ALG)  
2          Argentina (ARG)  
3            Armenia (ARM)  
4  Australasia (ANZ) [ANZ]

One nice feature of pandas is that it has the option to do **multi-level indexing**

In [2]:
#Loading Census dataset
df=pd.read_csv('census.csv')
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
0        4779736            4780127          4785161  ...          0.002295   
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
0         -0.193196          0.381066          0.582002         -0.467369   
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
0     1.030015     0.826644     1.383282     1.724718     0.712594  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  

[5 rows x 100 columns]

I want to see a list of all the unique values in a given column.

In [3]:
#Unique values present in SUMLEV column
df['SUMLEV'].unique()

array([40, 50])

In [5]:
#Retaining data only that has 'SUMLEV' equal to 50
df=df[df['SUMLEV']==50]
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   
5      50       3         6      1       9  Alabama   Blount County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   
5          57322              57322            57373  ...          1.807375   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   
5         -1.177622         -1.748766         -2.062535         -1.369970   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  
5     1.859511    -0.848580    -1.402476    -1.577232    -0.884411  

[5 rows x 100 columns]

In [6]:
#Getting rows that are summaries of state level and just keep county data
#Selected columns to be used
columns_to_keep = ['STNAME',
                   'CTYNAME',
                   'BIRTHS2010',
                   'BIRTHS2011',
                   'BIRTHS2012',
                   'BIRTHS2013',
                   'BIRTHS2014',
                   'BIRTHS2015',
                   'POPESTIMATE2010',
                   'POPESTIMATE2011',
                   'POPESTIMATE2012',
                   'POPESTIMATE2013',
                   'POPESTIMATE2014',
                   'POPESTIMATE2015']
df=df[columns_to_keep]
df.head()

STNAME         CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
1  Alabama  Autauga County         151         636         615         574   
2  Alabama  Baldwin County         517        2187        2092        2160   
3  Alabama  Barbour County          70         335         300         283   
4  Alabama     Bibb County          44         266         245         259   
5  Alabama   Blount County         183         744         710         646   

   BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
1         623         600            54660            55253            55175   
2        2186        2240           183193           186659           190396   
3         260         269            27341            27226            27159   
4         247         253            22861            22733            22642   
5         618         603            57373            57711            57776   

   POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
1            55038            55290            55347  
2           195126           199713           203709  
3            26973            26815            26489  
4            22512            22549            22583  
5            57734            57658            57673

In [7]:
#Having Dual index
df=df.set_index(['STNAME','CTYNAME'])
df

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                             
Alabama Autauga County            151         636         615         574   
        Baldwin County            517        2187        2092        2160   
        Barbour County             70         335         300         283   
        Bibb County                44         266         245         259   
        Blount County             183         744         710         646   
...                               ...         ...         ...         ...   
Wyoming Sweetwater County         167         640         595         657   
        Teton County               76         259         230         261   
        Uinta County               73         324         311         316   
        Washakie County            26         108          90          95   
        Weston County              26          81          74          93   

                           BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                      
Alabama Autauga County            623         600            54660   
        Baldwin County           2186        2240           183193   
        Barbour County            260         269            27341   
        Bibb County               247         253            22861   
        Blount County             618         603            57373   
...                               ...         ...              ...   
Wyoming Sweetwater County         629         620            43593   
        Teton County              249         269            21297   
        Uinta County              316         316            21102   
        Washakie County            96          90             8545   
        Weston County              77          79             7181   

                           POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                                
Alabama Autauga County               55253            55175            55038   
        Baldwin County              186659           190396           195126   
        Barbour County               27226            27159            26973   
        Bibb County                  22733            22642            22512   
        Blount County                57711            57776            57734   
...                                    ...              ...              ...   
Wyoming Sweetwater County            44041            45104            45162   
        Teton County                 21482            21697            22347   
        Uinta County                 20912            20989            21022   
        Washakie County               8469             8443             8443   
        Weston County                 7114             7065             7160   

                           POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                              
Alabama Autauga County               55290            55347  
        Baldwin County              199713           203709  
        Barbour County               26815            26489  
        Bibb County                  22549            22583  
        Blount County                57658            57673  
...                                    ...              ...  
Wyoming Sweetwater County            44925            44626  
        Teton County                 22905            23125  
        Uinta County                 20903            20822  
        Washakie County               8316             8328  
        Weston County                 7185             7234  

[3142 rows x 12 columns]

When you use a MultiIndex, you must provide the arguments in order by the level you wish to query. **Inside of the index, each column is called a level and the outermost column is level zero** 

In [8]:
#Now to see population results for a county, first we give the state followed by county
df.loc['Michigan','Washtenaw County']

BIRTHS2010            977
BIRTHS2011           3826
BIRTHS2012           3780
BIRTHS2013           3662
BIRTHS2014           3683
BIRTHS2015           3709
POPESTIMATE2010    345563
POPESTIMATE2011    349048
POPESTIMATE2012    351213
POPESTIMATE2013    354289
POPESTIMATE2014    357029
POPESTIMATE2015    358880
Name: (Michigan, Washtenaw County), dtype: int64

In [9]:
#To compare population of two different county of same states
df.loc[ [('Michigan', 'Washtenaw County'),
         ('Michigan', 'Wayne County')] ]

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME   CTYNAME                                                            
Michigan Washtenaw County         977        3826        3780        3662   
         Wayne County            5918       23819       23270       23377   

                           BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME   CTYNAME                                                     
Michigan Washtenaw County        3683        3709           345563   
         Wayne County           23607       23586          1815199   

                           POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME   CTYNAME                                                               
Michigan Washtenaw County           349048           351213           354289   
         Wayne County              1801273          1792514          1775713   

                           POPESTIMATE2014  POPESTIMATE2015  
STNAME   CTYNAME                                             
Michigan Washtenaw County           357029           358880  
         Wayne County              1766008          1759335

## Missing Values

We've seen a preview of how Pandas handles missing values using the None type and NumPy NaN values. Missing values are pretty common in data cleaning activities. There are couple of caveats and discussion points which we should address.

One of the handy functions that Pandas has for working with missing values is the filling function, fillna. This function takes a number or parameters, for instance, you could pass in a single value which is called a scalar value to change all of the missing data to one value. This isn't really applicable in this case, but it's a pretty common use case. Next up though is the method parameter. The two common fill values are ffill and bfill. ffill is for forward filling and it updates an na value for a particular cell with the value from the previous row. It's important to note that your data needs to be sorted in order for this to have the effect you might want. Data that comes from traditional database management systems usually has no order guarantee, just like this data. So be careful.

In [17]:
df=pd.read_csv('log.csv')
df

time    user          video  playback position paused  volume
0   1469974424  cheryl     intro.html                  5  False    10.0
1   1469974454  cheryl     intro.html                  6    NaN     NaN
2   1469974544  cheryl     intro.html                  9    NaN     NaN
3   1469974574  cheryl     intro.html                 10    NaN     NaN
4   1469977514     bob     intro.html                  1    NaN     NaN
5   1469977544     bob     intro.html                  1    NaN     NaN
6   1469977574     bob     intro.html                  1    NaN     NaN
7   1469977604     bob     intro.html                  1    NaN     NaN
8   1469974604  cheryl     intro.html                 11    NaN     NaN
9   1469974694  cheryl     intro.html                 14    NaN     NaN
10  1469974724  cheryl     intro.html                 15    NaN     NaN
11  1469974454     sue  advanced.html                 24    NaN     NaN
12  1469974524     sue  advanced.html                 25    NaN     NaN
13  1469974424     sue  advanced.html                 23  False    10.0
14  1469974554     sue  advanced.html                 26    NaN     NaN
15  1469974624     sue  advanced.html                 27    NaN     NaN
16  1469974654     sue  advanced.html                 28    NaN     5.0
17  1469974724     sue  advanced.html                 29    NaN     NaN
18  1469974484  cheryl     intro.html                  7    NaN     NaN
19  1469974514  cheryl     intro.html                  8    NaN     NaN
20  1469974754     sue  advanced.html                 30    NaN     NaN
21  1469974824     sue  advanced.html                 31    NaN     NaN
22  1469974854     sue  advanced.html                 32    NaN     NaN
23  1469974924     sue  advanced.html                 33    NaN     NaN
24  1469977424     bob     intro.html                  1   True    10.0
25  1469977454     bob     intro.html                  1    NaN     NaN
26  1469977484     bob     intro.html                  1    NaN     NaN
27  1469977634     bob     intro.html                  1    NaN     NaN
28  1469977664     bob     intro.html                  1    NaN     NaN
29  1469974634  cheryl     intro.html                 12    NaN     NaN
30  1469974664  cheryl     intro.html                 13    NaN     NaN
31  1469977694     bob     intro.html                  1    NaN     NaN
32  1469977724     bob     intro.html                  1    NaN     NaN

In Pandas we can sort either by index or by values. Here we'll just promote the time stamp to an index then sort on the index.

In [15]:
df=df.set_index('time')
df=df.sort_index()
df

user          video  playback position paused  volume
time                                                               
1469974424  cheryl     intro.html                  5  False    10.0
1469974424     sue  advanced.html                 23  False    10.0
1469974454  cheryl     intro.html                  6    NaN     NaN
1469974454     sue  advanced.html                 24    NaN     NaN
1469974484  cheryl     intro.html                  7    NaN     NaN
1469974514  cheryl     intro.html                  8    NaN     NaN
1469974524     sue  advanced.html                 25    NaN     NaN
1469974544  cheryl     intro.html                  9    NaN     NaN
1469974554     sue  advanced.html                 26    NaN     NaN
1469974574  cheryl     intro.html                 10    NaN     NaN
1469974604  cheryl     intro.html                 11    NaN     NaN
1469974624     sue  advanced.html                 27    NaN     NaN
1469974634  cheryl     intro.html                 12    NaN     NaN
1469974654     sue  advanced.html                 28    NaN     5.0
1469974664  cheryl     intro.html                 13    NaN     NaN
1469974694  cheryl     intro.html                 14    NaN     NaN
1469974724  cheryl     intro.html                 15    NaN     NaN
1469974724     sue  advanced.html                 29    NaN     NaN
1469974754     sue  advanced.html                 30    NaN     NaN
1469974824     sue  advanced.html                 31    NaN     NaN
1469974854     sue  advanced.html                 32    NaN     NaN
1469974924     sue  advanced.html                 33    NaN     NaN
1469977424     bob     intro.html                  1   True    10.0
1469977454     bob     intro.html                  1    NaN     NaN
1469977484     bob     intro.html                  1    NaN     NaN
1469977514     bob     intro.html                  1    NaN     NaN
1469977544     bob     intro.html                  1    NaN     NaN
1469977574     bob     intro.html                  1    NaN     NaN
1469977604     bob     intro.html                  1    NaN     NaN
1469977634     bob     intro.html                  1    NaN     NaN
1469977664     bob     intro.html                  1    NaN     NaN
1469977694     bob     intro.html                  1    NaN     NaN
1469977724     bob     intro.html                  1    NaN     NaN

In [18]:
#use multi-level indexing with time and user
df=df.set_index(['time','user'])
df=df.sort_index()
df

video  playback position paused  volume
time       user                                                   
1469974424 cheryl     intro.html                  5  False    10.0
           sue     advanced.html                 23  False    10.0
1469974454 cheryl     intro.html                  6    NaN     NaN
           sue     advanced.html                 24    NaN     NaN
1469974484 cheryl     intro.html                  7    NaN     NaN
1469974514 cheryl     intro.html                  8    NaN     NaN
1469974524 sue     advanced.html                 25    NaN     NaN
1469974544 cheryl     intro.html                  9    NaN     NaN
1469974554 sue     advanced.html                 26    NaN     NaN
1469974574 cheryl     intro.html                 10    NaN     NaN
1469974604 cheryl     intro.html                 11    NaN     NaN
1469974624 sue     advanced.html                 27    NaN     NaN
1469974634 cheryl     intro.html                 12    NaN     NaN
1469974654 sue     advanced.html                 28    NaN     5.0
1469974664 cheryl     intro.html                 13    NaN     NaN
1469974694 cheryl     intro.html                 14    NaN     NaN
1469974724 cheryl     intro.html                 15    NaN     NaN
           sue     advanced.html                 29    NaN     NaN
1469974754 sue     advanced.html                 30    NaN     NaN
1469974824 sue     advanced.html                 31    NaN     NaN
1469974854 sue     advanced.html                 32    NaN     NaN
1469974924 sue     advanced.html                 33    NaN     NaN
1469977424 bob        intro.html                  1   True    10.0
1469977454 bob        intro.html                  1    NaN     NaN
1469977484 bob        intro.html                  1    NaN     NaN
1469977514 bob        intro.html                  1    NaN     NaN
1469977544 bob        intro.html                  1    NaN     NaN
1469977574 bob        intro.html                  1    NaN     NaN
1469977604 bob        intro.html                  1    NaN     NaN
1469977634 bob        intro.html                  1    NaN     NaN
1469977664 bob        intro.html                  1    NaN     NaN
1469977694 bob        intro.html                  1    NaN     NaN
1469977724 bob        intro.html                  1    NaN     NaN

It's sometimes useful to use forward filling, sometimes backwards filling, and sometimes useful to just use a single number. More recently, the Pandas team introduced a method of filling missing values with a series which is the same length as your DataFrame. This makes it easy to derive values which are missing if you have the underlying to do so. For instance, if you're dealing with receipts and you have a column for final price and a column for discount but are missing information from the original price column, you can fill this automatically using fillna.
